src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Utility.cs

```diff
@@ -94,613 +71,275 @@ private static bool FirstCharInUInt32IsAscii(uint value)
-            }
-            else
-            {
-                throw new PlatformNotSupportedException();
+                return Test128(value.GetLower() | value.GetUpper());
             }
 
-            goto Finish;
-
-        FoundNonAsciiDataInCurrentDWord:
-
-            uint currentDWord;
-            Debug.Assert(!AllBytesInUInt32AreAscii(currentDWord), "Shouldn't be here unless we see non-ASCII data.");
-            pBuffer += CountNumberOfLeadingAsciiBytesFromUInt32WithSomeNonAsciiData(currentDWord);
-
-            goto Finish;
-
-        InputBufferLessThanOneVectorInLength:
-
-            // These code paths get hit if the original input length was less than one vector in size.
-            // We can't perform vectorized reads at this point, so we'll fall back to reading primitives
-            // directly. Note that all of these reads are unaligned.
-
-            Debug.Assert(bufferLength < SizeOfVector128);
-
-            // QWORD drain
-
-            if ((bufferLength & 8) != 0)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static bool Test512(Vector512<byte> value)
             {
-                if (UIntPtr.Size == sizeof(ulong))
-                {
-                    // If we can use 64-bit tzcnt to count the number of leading ASCII bytes, prefer it.
-
-                    ulong candidateUInt64 = Unsafe.ReadUnaligned<ulong>(pBuffer);
-                    if (!AllBytesInUInt64AreAscii(candidateUInt64))
-                    {
-                        // Clear everything but the high bit of each byte, then tzcnt.
-                        // Remember to divide by 8 at the end to convert bit count to byte count.
-
-                        candidateUInt64 &= UInt64HighBitsOnlyMask;
-                        pBuffer += (nuint)(BitOperations.TrailingZeroCount(candidateUInt64) >> 3);
-                        goto Finish;
-                    }
-                }
-                else
+                if (Avx512F.IsSupported)
                 {
-                    // If we can't use 64-bit tzcnt, no worries. We'll just do 2x 32-bit reads instead.
-
-                    currentDWord = Unsafe.ReadUnaligned<uint>(pBuffer);
-                    uint nextDWord = Unsafe.ReadUnaligned<uint>(pBuffer + 4);
-
-                    if (!AllBytesInUInt32AreAscii(currentDWord | nextDWord))
-                    {
-                        // At least one of the values wasn't all-ASCII.
-                        // We need to figure out which one it was and stick it in the currentMask local.
-
-                        if (AllBytesInUInt32AreAscii(currentDWord))
-                        {
-                            currentDWord = nextDWord; // this one is the culprit
-                            pBuffer += 4;
-                        }
-
-                        goto FoundNonAsciiDataInCurrentDWord;
-                    }
+                    return value.ExtractMostSignificantBits() != 0;
```

Comment:
This looks to generally just be doing `x != Vector<T>.Zero`, which should be the preferred mechanism to test for "AnyMatches" so we get the expected codegen

Reply:
This leads to worse codegen and higher cost (as per https://uica.uops.info/) therefore I changed it to movemask. Previous implementation also use this approach.

Reply:
That's likely something that should have a bug logged around it if so.

In general we want to prefer the standard patterns, even if slightly worse, as it leads to overall more maintainable code. It also helps push any underlying issues getting fixed and ensuring all code gets improvements.

Reply:
Sorry, just to clarify, it is faster to perform a single `movemask` which extracts exactly the MSBs that indicate if bytes are ASCII or not. The `!= Vector<T>.Zero` alternative also requires us to `& mask` first:

With `vpmovmskb` (uiCA: CQA) https://bit.ly/3RWQDlB predicted TP is 40B/cycle on Ice Lake
```asm
G_M59320_IG08:  ;; offset=0x0090
       vmovups  ymm0, ymmword ptr [rax]
       vmovups  ymm1, ymmword ptr [rax+0x20]
       vpor     ymm2, ymm0, ymm1
       vpmovmskb r10d, ymm2
       test     r10d, r10d
       jne      SHORT G_M59320_IG10
       add      rax, 64
       cmp      rax, r8
       jbe      SHORT G_M59320_IG08
```
With `((vector1 | vector2) & mask) != Vector256<byte>.Zero` https://bit.ly/3zDjnJS predicted TP is ~35B/cycle on Ice lake
```asm
G_M59320_IG09:  ;; offset=0x0098
       vmovups  ymm0, ymmword ptr [rax]
       vmovups  ymm1, ymmword ptr [rax+0x20]
       vpor     ymm2, ymm0, ymm1
       ;; We only care about MSBs in ymm2, vpmovmskb already does that for us so vpand is wasted work
       vpand    ymm2, ymm2, ymm3
       vpmovmskb r10d, ymm2
       test     r10d, r10d
       jne      SHORT G_M59320_IG11
       add      rax, 64
       cmp      rax, r8
       jbe      SHORT G_M59320_IG09
```

Do you want me to submit an issue for JIT recognizing that movemask's operand is a `vec & vec(MSB)` and simplifying it?

Reply:
> The != Vector<T>.Zero alternative also requires us to & mask first:

There is no `vpand; vpmovmskb`, we emit a direct `vptest` + `setcc` (or `jcc`) against the value and target: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8ovLrl5hcAbhrEAzE1IMAwgwDeNBjqZLmSBsAgQANgwCyACgBqMMBmjlSADgA8wAJ4YYAPgYIAJQMALz+VggMAGQMdg5OriyqsNg+Ht5+VnQILnSBwQCEIbH2jlDO7l4+viwAWjBQEHLUunpM5IbGZpa2pdCkAKxI6dUBwWEMEdEl8VCDSEkpaVWZ2bn5DEUzZfMjfnUNTdq6iu2dJubWcWUDznv+QaHhkTHX0LekizCpMPdZOXlCsU3lAPvcDo1mgBfIA==

On .NET 9+ this can further take advantage of embedded masking and other optimizations (as well as hoisting considerations, etc).

Reply:
We prefer standard patterns specifically because they get better with time and will take advantage of newer hardware where possible. That's much harder to do with other patterns, especially across all potential targets.

Reply:
Sorry, this was my mistake and on clean example I can reproduce the codegen you have linked. Thank you.
```asm
G_M59320_IG09:  ;; offset=0x00A0
       vmovups  ymm0, ymmword ptr [rax]
       vmovups  ymm1, ymmword ptr [rax+0x20]
       vpor     ymm2, ymm0, ymm1
       vptest   ymm2, ymm3
       jne      SHORT G_M59320_IG11
       add      rax, 64
       cmp      rax, r8
       jbe      SHORT G_M59320_IG09
```